cocos/2d/CCSpriteFrameCache.cpp

```diff
@@ -772,4 +740,92 @@ bool SpriteFrameCache::reloadTexture(const std::string& plist)
     return true;
 }
 
+
+void SpriteFrameCache::PlistFramesCache::insertFrame(const std::string &plist, const std::string &frame, SpriteFrame *spriteFrame)
+{
+    _spriteFrames.insert(frame, spriteFrame);   //add SpriteFrame
+
+    _indexPlist2Frames[plist].insert(frame);    //insert index plist->[frameName]
+    _indexFrame2plist[frame] = plist;           //insert index frameName->plist
+}
+
+bool SpriteFrameCache::PlistFramesCache::isPlistUsed(const std::string &plist) const
+{
+    //plist loaded && not empty
+    auto it = _indexPlist2Frames.find(plist);
+    return it != _indexPlist2Frames.end() && !it->second.empty();
+}
+
+bool SpriteFrameCache::PlistFramesCache::eraseFrame(const std::string &frame)
+{
+    _spriteFrames.erase(frame);                             //drop SpriteFrame
+    auto itFrame = _indexFrame2plist.find(frame);
+    if (itFrame != _indexFrame2plist.end())
+    {
+        _indexPlist2Frames[itFrame->second].erase(frame);   //update index plist->[frameNames]
+        _indexFrame2plist.erase(itFrame);                   //update index frame->plist
+        return true;
+    }
+    return false;
+}
+
+bool SpriteFrameCache::PlistFramesCache::eraseFrames(const std::vector<std::string> &frames)
+{
+    if (frames.empty()) return false;
+
+    for (auto frame : frames)
+    {
+        _spriteFrames.erase(frame);                             //drop SpriteFrame
+        auto itPlist = _indexFrame2plist.find(frame);
+        if (itPlist != _indexFrame2plist.end()) {
+            _indexPlist2Frames[itPlist->second].erase(frame);   //update index plist->[frameNames]
+            _indexFrame2plist.erase(itPlist);                   //update index frameName->plist
+        }
+    }
+
+    return true;
+}
+
+bool SpriteFrameCache::PlistFramesCache::erasePlistIndex(const std::string &plist)
+{
+    auto it = _indexPlist2Frames.find(plist);
+    if (it == _indexPlist2Frames.end()) return false;
+
+    auto &frames = it->second;
+    for (auto f : frames)
+    {
+        // !!do not!! call `_spriteFrames.erase(f);` to erase SpriteFrame
+        // only erase index here
+        _indexFrame2plist.erase(f);                             //erase plist frame frameName->plist
```

Comment:
Why not erase sprite frame?

Reply:
method `removeSpriteFramesFromDictionary` removes `sprite frames` internally. to keep the logic untouched, only update index here. 

https://github.com/cocos2d/cocos2d-x/blob/b26e1bb08648c11a9482d0736a236aebc345008a/cocos/2d/CCSpriteFrameCache.cpp#L512-L529